freelist: Performance, replace shift() with pop()

subzey opened the pull request:
Array#pop() is known to be faster than Array#shift().
In this case there's no difference from which side of the "pool" array
the object is retrieved.

thefourtheye commented:
Normally when we talk about performance, we create a benchmark suite to demonstrate the gain. Read more about it here https://github.com/nodejs/io.js/tree/master/benchmark

mscdex commented:
I agree with @thefourtheye, it would also be good to see numbers from said benchmark, especially on the master and next branches (and maybe even next+1 if it's working yet).

Trott commented:
ref: https://github.com/nodejs/io.js/issues/569

trevnorris commented:
It will be difficult to test this directly since it's an internal, and a micro benchmark between the two operations would be bogus. How could it be exposed for testing?

thefourtheye commented:
@trevnorris We can use `--expose-internals` flag introduced in #848, right?

trevnorris commented:
@thefourtheye Ah yes. Completely forgot about that one. Thanks for the reminder.

subzey commented:
Thanks for your patience. Here's a benchmark commit (I hope, I did it right).
Before was about 1 200 000 to 1 500 000, after is about 3 400 000 on my win7 machine.

trevnorris commented:
@subzey If you can even post the code in a gist I can help you write up a benchmark if you need it.

subzey commented:
@trevnorris Does 3531e82 looks good?

trevnorris commented:
@subzey Other than some style nits (spacing 'n stuff) it looks good.

thefourtheye commented:
@subzey The commit log should follow the guidelines listed [here](https://github.com/nodejs/io.js/blob/master/CONTRIBUTING.md#step-3-commit). Please check and update the commits.

trevnorris commented:
if it's an issue can be easily enough taken care of before we land it. just keep the reference for future contributions. ;)

mscdex commented:
@subzey Which branch did you test on to get those numbers?

subzey commented:
@mscdex, Latest release.
First I ran the code requiring `freelist` builtin into the executable, then `../../lib/internal/freelist` from this pr branch.

subzey commented:
Looks like this microoptimization is useless now (#2176), feel free to reject this PR.

thefourtheye commented:
@subzey Don't worry about it. Its deprecated for the users. It is still used internally by http module.

brendanashworth commented:
@subzey could you update your PR with test modifications cause of https://github.com/nodejs/io.js/commit/fef87fee1de60c3d5db2652ee2004a6e78d112ff?

subzey commented:
@brendanashworth I'm sorry, should I just merge `master` to this PR and run tests before pushing, or something else?

Fishrock123 commented:
@subzey He's asking if you could rebase against `master` and force-push your changes up here. :)

subzey commented:
Force-pushed a rebased commit into this PR; fixed eslint whitespace errors

thefourtheye commented:
@subzey Can you share some performance numbers, with `master` and with this `change`, if you don't mind?

subzey commented:
@thefourtheye, Sure.
Master: 1718948 (avg of 3 runs)
Branch: 3518114 (avg of 3 runs)

thefourtheye commented:
@subzey You might want to change the corresponding test as well, `test/parallel/test-freelist.js`. This will break the test.

subzey commented:
@thefourtheye, you're right. I'm terribly sorry for breaking the test, I'll fix it asap

brendanashworth commented on benchmark/misc/freelist.js:
```diff
@@ -0,0 +1,35 @@
+
+
+  // First, alloc N items
+  for (i = 0; i < n; i++) {
```
This benchmark doesn't really handle the common case (and exits way too fast). Do you think you could change it so it allocates / frees / allocates in blocks of say 500, all the way up to `n`? We try to have benchmarks take about 5-10 seconds to execute so numbers don't jump around as much.

jasnell commented:
@subzey ... is this still something you'd like to pursue?

subzey commented:
Sorry, I thought this PR is long dead.

Here's the updated benchmark and the numbers.

Before: 5 837
After: 66 610

jasnell commented:
@nodejs/ctc ... thoughts on this one?

mscdex commented:
I'm not seeing those kinds of results. When I run the benchmark on my local desktop (even bumped up `n` a bit more to give time for any optimizations to happen) I see the same results before and after this PR's changes.

subzey commented:
Here what I was doing:
- Run the benchmark several times
- Edit lib/internals/freelist.js changing `pop` to `shift` (as it was before)
- Run the benchmark again

Actually, I was surprised with an order of magnitude in run time difference, but I couldn't figure out what I was doing wrong.

mscdex commented:
@subzey How long is the benchmark running for you? I'd say make sure `n` is chosen such that it takes at least ~20 seconds? If you're still seeing the same large difference, what kind of platform are you running this on (software and hardware)?

mscdex commented:
Also, are you sure you are testing your patch with the latest copy of the master branch? There have been v8 upgrades (major and minor) since you first opened the PR in July.

subzey commented:
@mscdex I was running the lastest benchmark on node v.4.
I've just updated to v.5 and ran the benchmark again. I've got just the same x10 difference in benchmark run time:
```
PS C:\Users\akhlynovskiy\Code\io.js> Measure-Command { node benchmark/misc/freelist n=2000000 }


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 37
Milliseconds      : 289
Ticks             : 372894915
TotalDays         : 0,000431591336805556
TotalHours        : 0,0103581920833333
TotalMinutes      : 0,621491525
TotalSeconds      : 37,2894915
TotalMilliseconds : 37289,4915



PS C:\Users\akhlynovskiy\Code\io.js> Measure-Command { node benchmark/misc/freelist n=2000000 }


Days              : 0
Hours             : 0
Minutes           : 6
Seconds           : 10
Milliseconds      : 463
Ticks             : 3704632371
TotalDays         : 0,00428776894791667
TotalHours        : 0,10290645475
TotalMinutes      : 6,174387285
TotalSeconds      : 370,4632371
TotalMilliseconds : 370463,2371
```

I'm using Node 5.0.0 x64 on Win7. CPU is “AMD Athlon 64 X2 5000+ 2.61 GHz”.

subzey commented:
Is there any info I can provide?

ofrobots commented:
@subzey Thanks for your patience. The proposed change LGTM. I left some comments with minor nits.

The commit log messages should to follow these guidelines [here](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#step-3-commit). We can take care of those when landing if you're unsure.

mscdex commented on benchmark/misc/freelist.js:
```diff
@@ -0,0 +1,38 @@
+'use strict';
+
+var common = require('../common.js');
+var FreeList = require('../../lib/internal/freelist').FreeList;
```
This should be `require('internal/freelist')` with a note at the top that `--expose-internals` is needed to run the benchmark. Until benchmarks get special "Flags" code comments that get parsed like tests have, flags will need to be explicitly specified on the command-line.

mscdex commented on benchmark/misc/freelist.js:
```diff
@@ -0,0 +1,38 @@
+
+function main(conf) {
+  var n = conf.n;
+  var poolSize = 500;
```
It might be better to at least use the freelist max size currently used in core by the http module: 1000.

mscdex commented on benchmark/misc/freelist.js:
```diff
@@ -0,0 +1,38 @@
+  var j;
+  var used = [];
+
+  // First, alloc 500 items
```
s/500/`poolSize`/

mscdex commented:
A few nits but otherwise LGTM

mscdex commented on benchmark/misc/freelist.js:
```diff
@@ -0,0 +1,38 @@
+  for (i = 0; i < n; i++){
+    // Return all the N to the pool
+    for (j = 0; j < poolSize; j++) {
+      list.free(used[i]);
```
s/`used[i]`/`used[j]`/

mscdex commented on lib/internal/freelist.js:
```diff
@@ -10,7 +10,7 @@ exports.FreeList = function(name, max, constructor) {
 exports.FreeList.prototype.alloc = function() {
-  return this.list.length ? this.list.shift() :
+  return this.list.length ? this.list.pop() :
                             this.constructor.apply(this, arguments);
```
This can also be changed to simply `this.constructor();` for an additional speed increase since `http` is the only module that uses `FreeList` and `http` just passes in a factory function with no parameters as its "constructor."

subzey commented:
@ofrobots, @mscdex. The nits were fixed, I've squashed all the changes into 1 commit and rebased it on top of master (as I was asked previously). And also tried to make a perfect commit message.

Please take a look.

> This can also be changed to simply this.constructor(); for an additional speed increase since http is the only module that uses FreeList and http just passes in a factory function with no parameters as its "constructor."

A...actually, I'd prefer to do it in a separate branch. :)

ofrobots commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/1384/

mscdex commented on benchmark/misc/freelist.js:
```diff
@@ -0,0 +1,38 @@
+  bench.start();
+
+  for (i = 0; i < n; i++){
+    // Return all the N to the pool
```
s/N/items/

subzey commented:
> `// Return all the N to the pool`
> s/N/items/

Fixed

ofrobots commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/1723/

jasnell commented:
LGTM

ofrobots commented:
LGTM. Landed as c647e87.
